helper/random/string_generator.go

```diff
@@ -66,12 +66,17 @@ func sortCharset(chars string) string {
 	return string(r)
 }
 
-// StringGenerator generats random strings from the provided charset & adhering to a set of rules. The set of rules
+// StringGenerator generates random strings from the provided charset & adhering to a set of rules. The set of rules
 // are things like CharsetRule which requires a certain number of characters from a sub-charset.
 type StringGenerator struct {
 	// Length of the string to generate.
 	Length int `mapstructure:"length" json:"length"`
 
+	// Prefix will be prepended to the generated random string. It is not included in the configured length.
+	// It is intended to be used to make the generated string identifiable as a secret, so that detection of passwords
+	// leaked to logging or version control systems can be automated.
+	Prefix string `mapstructure:"prefix" json:"prefix"`
```

Comment:
I'm wondering if we should restrict the prefix to a fixed set of characters? The reason I ask is because I could see it being used in SQL injection. This is [an example](https://github.com/hashicorp/vault/blob/main/plugins/database/mysql/mysql.go#L24) of where the password is currently string replaced. This doesn't use prepared statements with parameters today (though I think it should).

I understand the likelihood of this happening is low. The password policy and database secrets APIs are authenticated and often configured by a privileged Vault user. In any case, I think we should guard against this with sufficient input validation if possible. Let me know your thoughts.

Reply:
That is an interesting thought.

However, the password creation policy system is not specific to databases, nor do all databases use SQL, nor are all SQL dialects all that similar, so it could be tricky to come up with a list of characters that meet all use-cases, whilst avoiding user surprise. (e.g. a hypothetical user reaction: "I can't use this character in my LDAP password prefix ... to prevent SQL injection? LDAP doesn't have anything to do with SQL! What's going on?!")

Additionally, the only way I can see this being used for SQL injection, would be if an adversary had already gained privileged access to modify Vault configuration - at which point, there's a fair chance they can just directly modify the database secrets engine configuration too.

My inclination, therefore, is to not attempt to combat SQL injection within the general password creation policy subsystem.

Reply:
> However, the password creation policy system is not specific to databases, nor do all databases use SQL, nor are all SQL dialects all that similar, so it could be tricky to come up with a list of characters that meet all use-cases, whilst avoiding user surprise. (e.g. a hypothetical user reaction: "I can't use this character in my LDAP password prefix ... to prevent SQL injection? LDAP doesn't have anything to do with SQL! What's going on?!")

I definitely hear you. Restricting the password policy prefix isn't the right way to solve the underlying issue. I'm wondering if there is something pragmatic we can do to support some prefixes while we address it. Starting with a limited character set still allows us to expand (for when that LDAP user asks :)) in the future. This could allow us to deliver _some_ value today while being able to expand the character set later on.

Do you think that's a reasonable compromise in the short term?

Reply:
Apologies for going silent on this conversation for many months - I have parted ways with the employer who had the original motivating use-case, so I no longer personally expect to make use of this feature any time soon.

As to the design issue being discussed ... I still feel that any character set restriction enforced at this point in the code, would provide insufficient practical benefit, for substantial downsides in extra complexity, documentation needed, and potential user confusion.

* I don't want to put a restriction on the whole password policy subsystem, which is used by many secrets engines, just for the database secrets engine.
* Especially when correctly defining the intersection of safe character sets for every current and future database plugin, is going to be an imperfect compromise anyway.
* And the entire stack of compromises is only meaningful at all, in defending against an adversary who has gained permission to write to `sys/policies/password/*`, but not `sys/policies/acl/*` or `database/config/*` - which, whilst admittedly a possible configuration of Vault, seems very niche indeed.

With those doubts, on top of no longer having a current personal motivation to progress this particular feature, I think it's best if I just mark this PR as draft for now - and, if you would prefer instead to close it, I'm OK with that.